local-volume/provisioner/pkg/deleter/blockcleaner.go

```diff
@@ -0,0 +1,247 @@
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+)
+
+const (
+	// ZeroDev is refers to zero device.
+	ZeroDev = "/dev/zero"
+	// DefaultBlockSize is the default block size to use in dd
+	DefaultBlockSize = "4096"
+	// ShredCommand the shred cli.
+	ShredCommand = "shred"
+	// MkfsCommand is the mkfs executable.
+	MkfsCommand = "mkfs"
+	// DDCommand is the dd executable.
+	DDCommand = "dd"
+	// WipeComand is the dd executable.
+	WipeComand = "wipefs"
+	// ShellCommand is the command to invoke shell
+	ShellCommand = "sh"
+
+	// DDFullErrExpected Error ouptput expected when running dd to fill a device
+	DDFullErrExpected = "No space left on device"
+	// NoErrExpected No error expected from running command
+	NoErrExpected = ""
+)
+
+// BlockCleaner Interface for invoking async script to do cleanup of block device.
+type BlockCleaner interface {
+	// DoCleanup deletes
+	DeleteBlockPV(pv *v1.PersistentVolume, mountPath string, config common.MountConfig) error
+	isRunning(pvName string) bool
+	markRunning(pvName string)
+	markDone(pvName string)
+}
+
+var _ BlockCleaner = &BlockCleanerImpl{}
+
+// BlockCleanerImpl Implementation of BLockCleaner interface
+type BlockCleanerImpl struct {
+	*common.RuntimeConfig
+	mutex     sync.Mutex
+	procTable map[string]time.Time
+}
+
+// NewBlockCleaner returns a BlockCleaner
+func NewBlockCleaner(config *common.RuntimeConfig) *BlockCleanerImpl {
+	return &BlockCleanerImpl{RuntimeConfig: config,
+		procTable: make(map[string]time.Time),
+	}
+}
+
+func (v *BlockCleanerImpl) isRunning(pvName string) bool {
+	v.mutex.Lock()
+	defer v.mutex.Unlock()
+	_, ok := v.procTable[pvName]
+	return ok
+}
+
+func (v *BlockCleanerImpl) markRunning(pvName string) {
+	v.mutex.Lock()
+	defer v.mutex.Unlock()
+	v.procTable[pvName] = time.Now()
+}
+
+func (v *BlockCleanerImpl) markDone(pvName string) {
+	v.mutex.Lock()
+	defer v.mutex.Unlock()
+	delete(v.procTable, pvName)
+}
+
+// DeleteBlockPV is meant to execute cleanup concurrently with other such cleanup processes.
+func (v *BlockCleanerImpl) DeleteBlockPV(pv *v1.PersistentVolume, mountPath string, config common.MountConfig) error {
+	iters, expErr, exe, args, err := v.getBlockDeleteCommand(mountPath, config)
+	if err != nil {
+		return err
+	}
+
+	if v.isRunning(pv.Name) {
+		// Run in progress, nothing to do,
+		return nil
+	}
+	v.markRunning(pv.Name)
+	go v.asyncDeleteBlockPV(pv, iters, expErr, exe, args...)
+
+	return nil
+}
+
+func (v *BlockCleanerImpl) getBlockDeleteCommand(mountPath string, config common.MountConfig) (int, string, string,
+	[]string, error) {
+	iters := config.BlockConf.Iterations
+	if config.BlockConf.Iterations <= 0 {
+		// Atleast one iteration must be run.
+		iters = 1
+	}
+
+	switch config.BlockConf.Cleaner {
+
+	case common.BlockCleanupFastRest, "":
+		// mkfs -F /dev/sda && wipefs -a /dev/sda
+		mkfsCmd := fmt.Sprintf("%s -F %q && %s -a %q", MkfsCommand, mountPath, WipeComand, mountPath)
+		return 1, NoErrExpected, ShellCommand, []string{"-c", mkfsCmd}, nil
```

Comment:
Might be worth a comment that executing this command multiple times will result in the same output, therefore iters = 1.

Reply:
code removed.